org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java

```diff
@@ -0,0 +1,67 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Konrad Windszus
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *      Konrad Windszus
+ *******************************************************************************/
+
+package org.eclipse.m2e.core.ui.archetype;
+
+import java.util.List;
+import java.util.Map;
+
+import org.osgi.annotation.versioning.ProviderType;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+
+import org.apache.maven.archetype.catalog.Archetype;
+import org.apache.maven.archetype.catalog.ArchetypeCatalog;
+import org.apache.maven.archetype.metadata.RequiredProperty;
+
+import org.eclipse.m2e.core.project.IArchetype;
+
+
+/**
+ * Provides operations to manage the catalog.
+ *
+ * @since 2.1.0 (package version 1.0.0)
+ */
+@ProviderType
+public interface ArchetypeManager {
+
+  /**
+   * Gets the required properties of an {@link IArchetype}.
+   *
+   * @param archetype the archetype possibly declaring required properties
+   * @param remoteArchetypeRepository the remote archetype repository, can be null
+   * @param monitor the progress monitor, can be null
+   * @return the required properties of the archetypes, null if none is found
```

Comment:
when your signature has a list, isn't it eclipse code guidelines to not return null and return an empty list instead?

Reply:
+1, it's generally a good practice to not return null for Arrays or Collections but empty ones; not only for Eclipse but for Java in general.

Reply:
Better return an `Optional<...>` instead then. There is a difference between an Archetype not found and having no required properties... one might even throw a CoreException if not found here ...

Reply:
> Better return an Optional<...> instead then.

I don't think so.

>  There is a difference between an Archetype not found and having no required properties..

Maybe, but that's not the goal of this method to detect that an Archetype is not found.  But your comment here emphasize that documentation is ambiguous and that there are different ways to interpret `none`.

> one might even throw a CoreException if not found here ...

CoreException is just an abstract exception in Eclipse world; creating a custom exception type would be better for an API.

Reply:
> > Better return an Optional<...> instead then.
> 
> I don't think so.

Why? Either is is optional (because it is fine / could happen) but then it is not valid to just return an empty list if nothing is found as a user of that method then falsely assumes everything is fine and one could potentially call this archetype (what will just fail because it is not found)...

> Maybe, but that's not the goal of this method to detect that an Archetype is not found. But your comment here emphasize that documentation is ambiguous and that there are different ways to interpret none.

Obviously there is a way to detect this, so either it is an error (= exception must be thrown) or a user has to be able to distinguish the cases.